### PR TITLE
feat(36327): libdef args rest type should always be array of any

### LINKED ISSUE #36327: Lib: args rest type should always be array of any
**TypeScript Version:**  3.8.0-dev.20200119

**Search Terms:** rest, ConstructorParameters, ReturnType
**Code**

```ts
/**
 * Obtain the parameters of a constructor function type in a tuple
 */
type ConstructorParameters<T extends new (...args: any) => any> = T extends new (...args: infer P) => any ? P : never;

```

**Expected behavior:**
`...args` rest param should extend `any[]`

**Actual behavior:**
`...args` has `any` type

**Playground Link:** 

**Related Issues:**

### PR DESCRIPTION
**Pull request by antongolub**
Fixes #36327 36327

### CODE REVIEW

**Comment by antongolub** [comment]
@sandersn 

My issue was related to another reason, so I closed it. But I still suggest describe `...args` rest param as `any[]` for consistency.

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -1473,22 +1473,22 @@ type NonNullable<T> = T extends null | undefined ? never : T;
...
  * Obtain the parameters of a constructor function type in a tuple
  */
-type ConstructorParameters<T extends new (...args: any) => any> = T extends new (...args: infer P) => any ? P : never;
+type ConstructorParameters<T extends new (...args: any[]) => any> = T extends new (...args: infer P) => any ? P : never;
 
 /**
  * Obtain the return type of a function type
  */
-type ReturnType<T extends (...args: any) => any> = T extends (...args: any) => infer R ? R : any;
+type ReturnType<T extends (...args: any[]) => any> = T extends (...args: any[]) => infer R ? R : any;
```
Comment (sandersn) [inline src/lib/es5.d.ts]:
```suggestion
type ReturnType<T extends (...args: any[]) => any> = T extends (...args: any) => infer R ? R : any;

```

File: src/lib/es5.d.ts [hunk 2]
```diff
@@ -1473,22 +1473,22 @@ type NonNullable<T> = T extends null | undefined ? never : T;
...
 
 /**
  * Obtain the return type of a constructor function type
  */
-type InstanceType<T extends new (...args: any) => any> = T extends new (...args: any) => infer R ? R : any;
+type InstanceType<T extends new (...args: any[]) => any> = T extends new (...args: any[]) => infer R ? R : any;
```
Comment (sandersn) [inline src/lib/es5.d.ts]:
```suggestion
type InstanceType<T extends new (...args: any[]) => any> = T extends new (...args: any) => infer R ? R : any;

```

**Comment by sandersn** [comment]
@typescript-bot run dt

**Comment by DanielRosenwasser** [comment]
Is this change going to handle `ReadonlyArray` and the like correctly? Does it exhibit any poor behavior when inferring from two overloads, one which is a subtype of `any[]` and one which is only assignable to `any[]`?

**Comment by antongolub** [comment]
@sandersn 
Done.

@DanielRosenwasser
I'm afraid I can’t clarify this. My concern was just about custom `IConstructable` interfaces to strictly comply `InstanceType` internal constraint.

**Comment by sandersn** [comment]
@antongolub can you add tests for the cases @DanielRosenwasser pointed out?
@DanielRosenwasser mind giving examples, particularly of the two-overload case?

**Comment by sandersn** [comment]
@typescript-bot run dt

maybe it will work this time

**Comment by sandersn** [comment]
@typescript-bot @weswigham DT is (almost) clean but it did *not* contribute to the PR's status checks that I can see.

**Comment by weswigham** [comment]
Uhhh, yeah, the whole "contributes to status checks" bit was broken when we swapped to using the merge commit a few weeks ago instead of the branch head, because devops is bad. Should we swap back? IMO, forcing manual merges is probably worth the correct status linkage?

**Comment by sandersn** [comment]
Not contributing to branch status is a feature rather than a bug given how broken our suite 2 tests usually are. As long as it's a known feature I'm fine with it, and I think it's better to keep perf testing the correct commit.

**Comment by sandersn** [comment]
@DanielRosenwasser filed #37193 with 2 examples that behave strangely with the *current* definitions. We need to figure out why they behave badly before moving forward with this PR. (Checking the new definitions would be interesting too, but I suspect that they behave the same.)

**Comment by sandersn** [comment]
This PR hasn't seen any activity for quite a while, so I'm going to close it to keep the number of open PRs manageable. Feel free to open a fresh PR or continue the discussion here.

**Pull request closed without merging**